### ISSUE TITLE
The color-contrast check doesn't work in JSDOM

### ISSUE THREAD

**Issue by thecristen** [dequelabs/axe-core#595]
Hi there, thanks for building this fantastic library!

I've never gotten bad color contrast to create violations when testing React components with JSDOM. It seems to always return a false passing result (well, when it's not timing out, but that's a known JSDOM issue). I've created an example based off of this project's jest+react example [here](https://github.com/thecristen/axe-core/tree/develop/doc/examples/jest%2Breact).

I'm guessing it doesn't work because [JSDOM has not implemented ranges as part of its DOM implementation](https://github.com/tmpvar/jsdom/issues/317), and the [code that checks for color contrast](https://github.com/dequelabs/axe-core/blob/develop/lib/rules/color-contrast-matches.js#L70) depends on that (`document.createRange()`). But the false passing result is a bit concerning.

**Comment by dylanb** [comment]
@thecristen what do you mean by "false passing result"? Is it returning this in the `violations` array or in the `incomplete` array?

**Comment by marcysutton** [comment]
FYI it wouldn't ever work because JSDOM doesn't support the necessary APIs like document.createRange. It's best to turn that rule off in JSDOM tests and run a page-level integration test with all the rules enabled instead.

**Comment by marcysutton** [comment]
I would just turn off the color contrast rule in JSDOM, because you're right that it won't ever work. We don't recommend testing color contrast in JSDOM due to the lack of support.

**Comment by dylanb** [comment]
@marcysutton yes, but it should be passing the results for this rule back in the `incomplete` array - no?

**Comment by thecristen** [comment]
Thanks for the comments, this made me look at it a bit carefully. I shouldn't have taken the absence of violations as a "passing" indication (though, I've seen too many tests follow that pattern..). @dylanb you're right in that the results end up in the `incomplete` array - that's great!

**Comment by kalpeshsingh** [comment]
> Thanks for the comments, this made me look at it a bit carefully. I shouldn't have taken the absence of violations as a "passing" indication (though, I've seen too many tests follow that pattern..). @dylanb you're right in that the results end up in the `incomplete` array - that's great!

Curious to know what was your next step here? Did you remove this from unit testing as it requires manual intervention?

**Comment by mohanraj-r** [comment]
@dylanb @marcysutton looks like the range and selection APIs were added to JSDOM in v16 (https://github.com/jsdom/jsdom/pull/2719). Does this mean color contrast checks can work with axe-core in JSDOM? Does it need any updates to axe-core or should it work as is with JSDOM v16 or higher?

**Comment by straker** [comment]
@mohanraj-r I tested out jsdom 18.0.1. Looks like something about the range API is still not fully working as I get this error when running a test:

```
"message": "`TypeError: range.getClientRects is not a function` - feature unsupported in your environment. Skipping color-contrast rule.",
        "ruleId": "color-contrast",
        "stack": "Error\n    at new SupportError (/Users/stevenlambert/deque/axe-core/doc/examples/jsdom/node_modules/axe-core/axe.js:52:18)\n    at Rule.run2 [as run] (/Users/stevenlambert/deque/axe-core/doc/examples/jsdom/node_modules/axe-core/axe.js:19961:16)\n    at /Users/stevenlambert/deque/axe-core/doc/examples/jsdom/node_modules/axe-core/axe.js:20719:15\n    at pop (/Users/stevenlambert/deque/axe-core/doc/examples/jsdom/node_modules/axe-core/axe.js:6649:18)\n    at Object.defer (/Users/stevenlambert/deque/axe-core/doc/examples/jsdom/node_modules/axe-core/axe.js:6671:11)\n
```

**Comment by paschalidi** [comment]
Seeing the same here. Is there any known fix for this?

**Comment by straker** [comment]
Unfortunately without full support of the APIs we use, there's not much we can do for a fix. I just tested with jsom 19 and still am getting the not supported message.

```
`TypeError: range.getClientRects is not a function` - feature unsupported in your environment. Skipping color-contrast rule
``

**Comment by tounsoo** [comment]
Is this still the case? Or does anyone know another alternative library that will check for color contrast?

**Comment by ashleyfigure** [comment]
> 

@tounsoo  I'm a little late to the party but we've been using axe-core with Cypress to test color contrast. Cypress runs tests in a browser vs JSDOM.
